### PR TITLE
[Draft]Use ValueStringBuilder in Encode/DecodeName

### PR DESCRIPTION
**Pull request by TrayanZapryanov**
Follow up of [Question](https://github.com/dotnet/runtime/issues/65884)

Results:
Before

|                          Method |         Mean |      Error |     StdDev |       Median |          Min |          Max |  Gen 0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
|                      EncodeName |    81.036 ns |  0.9210 ns |  0.7690 ns |    80.738 ns |    80.248 ns |    82.882 ns | 0.0267 |     224 B |
|               EncodeName_Larger |    93.642 ns |  1.4571 ns |  1.3629 ns |    93.391 ns |    89.993 ns |    95.955 ns |      - |         - |
|              EncodeName_Invalid |   188.668 ns |  1.7830 ns |  1.4889 ns |   189.065 ns |   183.973 ns |   189.956 ns | 0.0523 |     440 B |
|       EncodeName_Invalid_Larger | 1,569.636 ns | 19.6314 ns | 19.2807 ns | 1,564.548 ns | 1,529.487 ns | 1,615.296 ns | 0.4290 |    3608 B |
|                      DecodeName |     5.210 ns |  0.0916 ns |  0.0857 ns |     5.219 ns |     5.042 ns |     5.348 ns |      - |         - |
|               DecodeName_Larger |     6.187 ns |  0.0345 ns |  0.0306 ns |     6.188 ns |     6.145 ns |     6.259 ns |      - |         - |
|              DecodeName_Invalid |   924.028 ns |  4.9796 ns |  4.6579 ns |   924.000 ns |   917.484 ns |   930.695 ns | 0.2408 |    2040 B |
|       DecodeName_Invalid_Larger | 1,048.205 ns |  7.7555 ns |  7.2545 ns | 1,047.681 ns | 1,037.933 ns | 1,061.621 ns | 0.3364 |    2840 B |

After
|                          Method |         Mean |      Error |     StdDev |       Median |          Min |          Max |  Gen 0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
|                      EncodeName |    83.026 ns |  0.1343 ns |  0.1121 ns |    83.042 ns |    82.807 ns |    83.222 ns | 0.0092 |      80 B |
|               EncodeName_Larger |    78.872 ns |  0.3150 ns |  0.2947 ns |    78.913 ns |    78.384 ns |    79.302 ns |      - |         - |
|              EncodeName_Invalid |   145.458 ns |  2.2386 ns |  1.9845 ns |   144.930 ns |   141.253 ns |   148.942 ns | 0.0158 |     136 B |
|       EncodeName_Invalid_Larger | 1,425.675 ns | 15.9157 ns | 13.2903 ns | 1,423.797 ns | 1,411.260 ns | 1,458.079 ns | 0.2784 |    2360 B |
|                      DecodeName |     6.114 ns |  0.1560 ns |  0.1383 ns |     6.104 ns |     5.802 ns |     6.358 ns |      - |         - |
|               DecodeName_Larger |     7.467 ns |  0.0599 ns |  0.0531 ns |     7.449 ns |     7.389 ns |     7.566 ns |      - |         - |
|              DecodeName_Invalid |   908.402 ns |  6.5217 ns |  5.4459 ns |   908.704 ns |   899.138 ns |   916.882 ns | 0.2165 |    1816 B |
|       DecodeName_Invalid_Larger | 1,059.641 ns |  5.5400 ns |  4.9111 ns | 1,059.661 ns | 1,052.079 ns | 1,067.418 ns | 0.2631 |    2216 B |

### CODE REVIEW

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs [hunk 1]
```diff
@@ -107,6 +104,13 @@ public class XmlConvert
                 matchPos = m.Index;
             }
 
+            ValueStringBuilder bufBld =
+                matchPos == -1
+                    ? default
+                    : length < 50
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:110]: would be good to create constant for this, also I'd increase this to 128 or 256 which are the two common numbers we use in other places

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs [hunk 2]
```diff
@@ -216,29 +217,35 @@ public class XmlConvert
                 matchPos = m.Index - 1;
             }
 
+            ValueStringBuilder bufBld =
+                matchPos == -1
+                    ? default
+                    : length < 50
+                        ? new ValueStringBuilder(stackalloc char[length])
```
[the remark this answers is not in the document]
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs]: pushed changes with constant

[the remark this answers is not in the document]
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs]: @gfoidl Unfortunately this doesn't help at least on my PC benchmark.
Still I can see regression for correct inputs.
You can check latest results.

**Comment by TrayanZapryanov** [comment]
@krwq , @gfoidl I have two more ideas for improvement here:
1) As regex was used, source generator can be used. But I have no idea how this should be done.
2) Based on your comment that we should optimize only "happy" path, then we can use new method of regex for counting if there are matches in the fast path. If there are then go to slow path like the code is now.

Any thoughts on this?

**Comment by TrayanZapryanov** [comment]
> @krwq , @gfoidl I have two more ideas for improvement here:
> 
> 1. As regex was used, source generator can be used. But I have no idea how this should be done.
> 2. Based on your comment that we should optimize only "happy" path, then we can use new method of regex for counting if there are matches in the fast path. If there are then go to slow path like the code is now.
> 
> Any thoughts on this?

Please ignore my comment. I debugged EncodeName method and allocations(80bytes which left) are not from regex usage.
Another thing that I've found is that my benchmark is also not correct as baseline of it should be string even without spaces.
In case there are no spaces - allocations are zero:
`[Benchmark]
		public string EncodeName() => XmlConvert.EncodeName("notencodedname");

		[Benchmark]
		public string EncodeName_with_space() => XmlConvert.EncodeName("not encoded name");` 


These are my new results - comparing original/ValueStringBuilder/StringBuilderCache.
Looks like we can achieve better memory - only changing new StringBuilder() with StringBuilderCache.Accuire()

Before
|                          Method |         Mean |      Error |     StdDev |       Median |          Min |          Max |  Gen 0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
|                      EncodeName |    22.460 ns |  0.0689 ns |  0.0611 ns |    22.464 ns |    22.375 ns |    22.591 ns |      - |         - |
|           EncodeName_with_space |    80.833 ns |  0.2598 ns |  0.2430 ns |    80.785 ns |    80.458 ns |    81.357 ns | 0.0267 |     224 B |
|               EncodeName_Larger |    75.807 ns |  0.1174 ns |  0.0981 ns |    75.796 ns |    75.634 ns |    75.994 ns |      - |         - |
|              EncodeName_Invalid |   200.549 ns |  1.6427 ns |  1.3718 ns |   200.250 ns |   198.745 ns |   203.874 ns | 0.0523 |     440 B |
|       EncodeName_Invalid_Larger | 1,475.367 ns | 28.3762 ns | 30.3622 ns | 1,487.094 ns | 1,418.523 ns | 1,526.133 ns | 0.4287 |    3608 B |
|                      DecodeName |     4.065 ns |  0.0413 ns |  0.0345 ns |     4.070 ns |     3.987 ns |     4.116 ns |      - |         - |
|           DecodeName_with_space |   351.029 ns |  3.0696 ns |  2.8713 ns |   351.544 ns |   342.587 ns |   354.894 ns | 0.1062 |     896 B |
|               DecodeName_Larger |     7.377 ns |  0.1631 ns |  0.1525 ns |     7.423 ns |     7.026 ns |     7.635 ns |      - |         - |
|              DecodeName_Invalid |   857.166 ns |  4.5961 ns |  4.0743 ns |   858.121 ns |   846.551 ns |   862.404 ns | 0.2405 |    2040 B |
|       DecodeName_Invalid_Larger |   975.957 ns |  4.4592 ns |  3.9530 ns |   975.943 ns |   965.971 ns |   981.666 ns | 0.3361 |    2840 B |


After - ValueStringBuilder

|                          Method |         Mean |      Error |     StdDev |       Median |          Min |          Max |  Gen 0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
|                      EncodeName |    27.831 ns |  0.3200 ns |  0.2993 ns |    27.900 ns |    26.785 ns |    28.040 ns |      - |         - |
|           EncodeName_with_space |    89.707 ns |  0.3064 ns |  0.2867 ns |    89.704 ns |    89.224 ns |    90.163 ns | 0.0095 |      80 B |
|               EncodeName_Larger |    98.104 ns |  0.1849 ns |  0.1639 ns |    98.116 ns |    97.807 ns |    98.378 ns |      - |         - |
|              EncodeName_Invalid |   157.043 ns |  0.4825 ns |  0.4277 ns |   156.934 ns |   156.548 ns |   158.012 ns | 0.0158 |     136 B |
|       EncodeName_Invalid_Larger | 1,471.369 ns | 24.9686 ns | 22.1340 ns | 1,472.582 ns | 1,443.176 ns | 1,513.464 ns | 0.2812 |    2360 B |
|                      DecodeName |     5.275 ns |  0.0259 ns |  0.0242 ns |     5.281 ns |     5.238 ns |     5.317 ns |      - |         - |
|           DecodeName_with_space |   353.655 ns |  4.5728 ns |  4.0537 ns |   354.604 ns |   340.567 ns |   357.635 ns | 0.0866 |     728 B |
|               DecodeName_Larger |     7.555 ns |  0.0305 ns |  0.0254 ns |     7.569 ns |     7.510 ns |     7.584 ns |      - |         - |
|              DecodeName_Invalid |   923.422 ns |  5.6303 ns |  5.2666 ns |   920.817 ns |   917.580 ns |   934.043 ns | 0.2140 |    1816 B |
|       DecodeName_Invalid_Larger | 1,094.594 ns |  4.3834 ns |  4.1002 ns | 1,094.800 ns | 1,088.441 ns | 1,100.708 ns | 0.2612 |    2216 B |


After - StringBuilderCache

|                          Method |         Mean |      Error |     StdDev |       Median |          Min |          Max |  Gen 0 | Allocated |
|-------------------------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
|                      EncodeName |    20.954 ns |  0.1912 ns |  0.1788 ns |    20.973 ns |    20.563 ns |    21.244 ns |      - |         - |
|           EncodeName_with_space |    74.062 ns |  0.1876 ns |  0.1755 ns |    74.059 ns |    73.756 ns |    74.333 ns | 0.0095 |      80 B |
|               EncodeName_Larger |    74.622 ns |  0.7414 ns |  0.6572 ns |    74.800 ns |    72.410 ns |    75.017 ns |      - |         - |
|              EncodeName_Invalid |   139.483 ns |  2.5666 ns |  2.4008 ns |   138.537 ns |   137.183 ns |   143.600 ns | 0.0160 |     136 B |
|       EncodeName_Invalid_Larger | 1,499.118 ns | 20.9903 ns | 19.6344 ns | 1,490.569 ns | 1,473.490 ns | 1,534.689 ns | 0.4254 |    3608 B |
|                      DecodeName |     4.566 ns |  0.0271 ns |  0.0253 ns |     4.560 ns |     4.538 ns |     4.616 ns |      - |         - |
|           DecodeName_with_space |   359.849 ns |  2.3792 ns |  2.1091 ns |   359.678 ns |   355.096 ns |   363.083 ns | 0.0859 |     728 B |
|               DecodeName_Larger |     6.509 ns |  0.0405 ns |  0.0359 ns |     6.505 ns |     6.464 ns |     6.598 ns |      - |         - |
|              DecodeName_Invalid |   948.486 ns |  3.6826 ns |  3.2645 ns |   949.632 ns |   942.643 ns |   952.425 ns | 0.2160 |    1816 B |
|       DecodeName_Invalid_Larger | 1,048.151 ns |  5.3282 ns |  4.7233 ns | 1,046.998 ns | 1,040.618 ns | 1,057.597 ns | 0.2643 |    2216 B |

**Comment by TrayanZapryanov** [comment]
@stephentoub Just want to share some results after your change in System.Private.Xml with regex generator:
There are really good wins here. These are results with fresh build. Above are results without your changes:

|                          Method |         Mean |     Error |    StdDev |       Median |          Min |          Max |  Gen 0 |  Gen 1 | Allocated |
|-------------------------------- |-------------:|----------:|----------:|-------------:|-------------:|-------------:|-------:|-------:|----------:|
|                      EncodeName |    21.584 ns | 0.3147 ns | 0.2790 ns |    21.671 ns |    20.883 ns |    21.883 ns |      - |      - |         - |
|           EncodeName_with_space |    78.327 ns | 0.3589 ns | 0.3357 ns |    78.417 ns |    77.428 ns |    78.680 ns | 0.0266 |      - |     224 B |
|               EncodeName_Larger |    85.827 ns | 1.1513 ns | 0.9614 ns |    85.742 ns |    84.275 ns |    87.975 ns |      - |      - |         - |
|              EncodeName_Invalid |   197.543 ns | 1.6026 ns | 1.2512 ns |   197.328 ns |   195.828 ns |   200.864 ns | 0.0522 |      - |     440 B |
|       EncodeName_Invalid_Larger | 1,430.983 ns | 7.1841 ns | 6.7200 ns | 1,430.460 ns | 1,419.574 ns | 1,442.578 ns | 0.3998 |      - |    3368 B |
|                      DecodeName |     3.950 ns | 0.0267 ns | 0.0223 ns |     3.955 ns |     3.902 ns |     3.984 ns |      - |      - |         - |
|           DecodeName_with_space |   226.265 ns | 4.9205 ns | 5.6664 ns |   224.520 ns |   216.806 ns |   236.851 ns | 0.0975 |      - |     816 B |
|               DecodeName_Larger |     6.610 ns | 0.1300 ns | 0.1216 ns |     6.628 ns |     6.262 ns |     6.755 ns |      - |      - |         - |
|              DecodeName_Invalid |   524.920 ns | 1.2977 ns | 1.0837 ns |   524.660 ns |   523.517 ns |   527.225 ns | 0.2145 |      - |    1800 B |
|       DecodeName_Invalid_Larger |   621.897 ns | 2.5720 ns | 2.4058 ns |   621.385 ns |   618.965 ns |   626.436 ns | 0.3084 | 0.0026 |    2600 B |

**Comment by TrayanZapryanov** [comment]
@krwq , @gfoidl I will close this as it leads to regression also it improves not "happy" paths.
Is it okay ?

**Comment by gfoidl** [comment]
I wouldn't close the PR (now). Time is one measure, another one is memory allocated, and for the latter it shows in general an improvement.

Depending on how much effort you want to put into this, you could do a profile to see where / why ValueStringBuilder lacks on perf, so that these areas can be improved. Then all usages of the VSB could gain that benefit.

But I'll leave that decision up to @krwq (I'm not member of the team here).

**Pull request closed without merging**